feat: integrate Google Cloud Model Armor guardrails

colesmcintosh opened the pull request:
## Title

Integrate Google Cloud Model Armor guardrails with LiteLLM

## Relevant issues

Fixes LIT-298

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🆕 New Feature

## Changes

This PR integrates Google Cloud Model Armor API with LiteLLM to provide guardrails functionality similar to Bedrock Guards.

### Key Features:
- **Pre-call sanitization**: Sanitizes user prompts before sending to LLM using `sanitizeUserPrompt` endpoint
- **Post-call sanitization**: Sanitizes model responses using `sanitizeModelResponse` endpoint  
- **Vertex AI authentication**: Leverages existing `VertexBase` class for authentication
- **Content masking**: Supports masking sensitive content in both requests and responses
- **Streaming support**: Handles streaming responses with content sanitization

### Implementation Details:
- Created `ModelArmorGuardrail` class extending `CustomGuardrail` and `VertexBase`
- Added configuration model `ModelArmorConfigModel` with required parameters (template_id, project_id, location, etc.)
- Registered Model Armor in guardrail registry and initializers
- Comprehensive test suite covering all functionality

### Usage Example:
```yaml
guardrails:
  - guardrail_name: "model-armor-guard"
    litellm_params:
      guardrail: model_armor
      mode: ["pre_call", "post_call"]
      template_id: "your-template-id"
      project_id: "your-project-id"
      location: "us-central1"
      mask_request_content: true
      mask_response_content: true
```

### Test Results:
All tests passing:
```
tests/guardrails_tests/test_model_armor.py::test_model_armor_pre_call_hook_sanitization PASSED
tests/guardrails_tests/test_model_armor.py::test_model_armor_pre_call_hook_blocked PASSED
tests/guardrails_tests/test_model_armor.py::test_model_armor_post_call_hook_sanitization PASSED
tests/guardrails_tests/test_model_armor.py::test_model_armor_with_list_content PASSED
tests/guardrails_tests/test_model_armor.py::test_model_armor_api_error_handling PASSED
tests/guardrails_tests/test_model_armor.py::test_model_armor_credentials_handling PASSED
tests/guardrails_tests/test_model_armor.py::test_model_armor_streaming_response PASSED
```

### Notes:
- Model Armor API is accessed via REST endpoints, no separate SDK required
- Uses same authentication mechanisms as other Vertex AI services
- Provides flexible configuration for different deployment scenarios

colesmcintosh committed:
feat: integrate Google Cloud Model Armor guardrails (LIT-298)

- Add ModelArmorGuardrail class that extends CustomGuardrail and VertexBase
- Support for both pre-call (sanitizeUserPrompt) and post-call (sanitizeModelResponse) sanitization
- Integrate with existing Vertex AI authentication using VertexBase
- Add configuration model for Model Armor in guardrail types
- Register Model Armor in guardrail initializers and registry
- Include comprehensive test suite for Model Armor functionality
- Support content masking for both requests and responses
- Handle streaming responses with content sanitization

This integration allows LiteLLM to use Google Cloud Model Armor API for
content moderation and sanitization, providing similar functionality to
Bedrock Guards but using Google Cloud's security infrastructure.

colesmcintosh committed:
fix: remove unused imports flagged by ruff linter

- Remove unused asyncio import at top level (moved to local import where needed)
- Remove unused TextCompletionResponse import

colesmcintosh committed:
fix: remove additional unused imports

- Remove unused TextChoices import from line 34
- Remove duplicate asyncio import from line 384
- Replace asyncio.iscoroutine() with hasattr check for __await__

colesmcintosh committed:
fix: remove final unused imports

- Remove unused 'import sys' from line 9
- Remove unused 'StreamingChoices' from imports

colesmcintosh committed:
fix: remove unused import from model_armor.py

- Remove unused 'import os' from the top of the file

colesmcintosh committed:
fix: remove commented-out header from model_armor.py

- Eliminate unnecessary comments at the top of the file to improve code clarity.

ishaan-jaff reviewed (changes requested):
waiting on manual testing

colesmcintosh committed:
Merge upstream main into feature/model-armor-integration

colesmcintosh commented:
[UI Demo](https://www.loom.com/share/71c3fb41e2aa4658b30f754402d0331b?sid=f6fd3e39-7103-4837-a1b6-064e58724cf9)

colesmcintosh committed:
feat(guardrails): Add Model Armor UI support

- Convert model_armor.py to a directory structure with __init__.py for dynamic discovery
- Add get_config_model() method to ModelArmorGuardrail class for UI integration
- Add ui_friendly_name() to ModelArmorConfigModel returning "Google Cloud Model Armor"
- Remove manual registration from guardrail_registry.py to use dynamic discovery
- Model Armor now appears in the UI guardrails dropdown with proper configuration fields

This enables users to configure Model Armor guardrails through the LiteLLM UI interface.

colesmcintosh committed:
fix(guardrails): Fix undefined name 'GuardrailConfigModel' in Model Armor

- Import TYPE_CHECKING and GuardrailConfigModel from base module
- Fixes F821 linting error for undefined name in type annotation
- Follows same pattern as other guardrail implementations

colesmcintosh committed:
fix(guardrails): Fix Model Armor type errors and config model inheritance

- Create ModelArmorGuardrailConfigModel that properly inherits from GuardrailConfigModel base class
- Move config model to litellm/types/proxy/guardrails/guardrail_hooks/model_armor.py following convention
- Update get_config_model() to return the properly typed config model
- Remove ModelArmorConfigModel from LitellmParams inheritance chain
- Add template_id field to BaseLitellmParams instead

This fixes the mypy type errors and follows the same pattern as other guardrail implementations.

colesmcintosh committed:
fix(guardrails): Add missing Model Armor fields to BaseLitellmParams

- Add location, credentials, api_endpoint, and fail_on_error fields
- Fixes mypy errors about missing attributes in LitellmParams
- All Model Armor configuration parameters are now properly defined

krrishdholakia commented:
Hey @colesmcintosh - can you please include a screenshot / video of the guardrail being used in test key, to confirm it works? 

<img width="1113" height="267" alt="Screenshot 2025-07-14 at 10 37 39 PM" src="https://github.com/user-attachments/assets/2aadf27f-9ddf-4412-95a2-091457bacc4f" />

krrishdholakia commented on tests/guardrails_tests/test_model_armor.py:
move inside `test_litellm/` to run via github action

krrishdholakia commented on litellm/proxy/guardrails/guardrail_hooks/model_armor/model_armor.py:
```diff
@@ -0,0 +1,458 @@
+        else:
+            return {"model_response_data": {"text": content}}
+
+    def _extract_content_from_messages(self, messages: List[AllMessageValues]) -> str:
```
can we use the helper from `prompt_factory/common_utils` instead?

krrishdholakia commented on litellm/proxy/guardrails/guardrail_hooks/model_armor/model_armor.py:
```diff
@@ -0,0 +1,458 @@
+            return data
+
+        # Extract content from messages
+        content = self._extract_content_from_messages(messages)
```
shouldn't we just be extracting the last consecutive block of user messages? 

we risk exceeding context window limits otherwise (see azure/prompt_shield for reference)

krrishdholakia commented on litellm/proxy/guardrails/guardrail_hooks/model_armor/model_armor.py:
```diff
@@ -0,0 +1,458 @@
+
+        return "\n".join(content_parts)
+
+    def _extract_content_from_response(
```
what happens when response is not ModelResponse? (e.g. tts call)

colesmcintosh committed:
fix: Apply PR review feedback for Model Armor guardrail

- Move test file to tests/test_litellm/ for GitHub Actions
- Extract only last consecutive user messages to avoid context limits
- Use get_content_from_model_response helper for response extraction
- Handle non-ModelResponse types (e.g., TTS) gracefully
- Maintain newline separation for multi-part content

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/model-armor-integration

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/model-armor-integration

colesmcintosh committed:
refactor: Simplify message content extraction in ModelArmorGuardrail

- Removed the custom _extract_content_from_messages method.
- Integrated get_last_user_message helper for improved content extraction.
- Updated test to reflect changes in content formatting.

colesmcintosh committed:
refactor: Remove unused import in model_armor.py

- Deleted the unused import of AllMessageValues to clean up the codebase.

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/model-armor-integration

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/model-armor-integration

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/model-armor-integration

colesmcintosh committed:
Add unit tests for Model Armor guardrail functionality

- Implement tests for pre-call and post-call hooks, including content sanitization and blocking behavior.
- Validate error handling for API responses and credential management.
- Test streaming responses and handling of list content in user messages.
- Ensure proper assertions for API interactions and response sanitization.

colesmcintosh committed:
Add comprehensive test coverage for Model Armor guardrail

- Add test for requests with no messages field
- Add test for empty message content handling
- Add test for system/assistant-only messages
- Add test for fail_on_error=False behavior
- Add test for custom API endpoint configuration
- Add test for dictionary credentials (non-file path)
- Add test for action=NONE response handling
- Add test for missing sanitized_text field fallback
- Add test for non-text response types (TTS/image)
- Add test for auth token refresh behavior

All tests ensure robust edge case handling and proper error management.

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/model-armor-integration

colesmcintosh committed:
Merge branch 'BerriAI:main' into feature/model-armor-integration

colesmcintosh committed:
feat: improve Model Armor handling of non-ModelResponse types

- Add debug logging when skipping non-text responses (TTS, images, etc.)
- Improve docstring to clarify behavior for non-text responses
- Add test coverage for non-ModelResponse handling
- Ensure guardrail gracefully skips processing for response types it cannot handle

colesmcintosh commented:
[Model Guardrail Demo Video](https://www.loom.com/share/4ece881182d149edb8ef7a1f9313a550?sid=765bc286-1413-4926-8555-0768422512fb)

krrishdholakia commented on litellm/proxy/guardrails/guardrail_hooks/model_armor/__init__.py:
@colesmcintosh we should probably add this to docs as well, so a user knows this exists - https://docs.litellm.ai/docs/proxy/guardrails/quick_start
